integrations/gcp/gcp_core/search/utils.py

```diff
@@ -1,50 +1,63 @@
-import asyncio
-from typing import Any, Callable, AsyncGenerator
-from loguru import logger
-from port_ocean.core.ocean_types import ASYNC_GENERATOR_RESYNC_TYPE
-
-from google.api_core.retry_async import AsyncRetry
-from google.api_core.exceptions import (
-    TooManyRequests,
-    ServiceUnavailable,
-)
-
-
-RETRIABLE_ERROR_TYPES = (TooManyRequests, ServiceUnavailable)
-INITIAL_DELAY_BEFORE_FIRST_RETRY_ATTEMPT = 1.0
-MAXIMUM_DELAY_BETWEEN_RETRY_ATTEMPTS = 60.0
-MULTIPLIER_FOR_EXPONENTIAL_BACKOFF = 2.0
-TOTAL_TIME_ALLOWED_FOR_RETRIES = 300.0
-
-MAX_CONCURRENT_CALLS = 10
-semaphore = asyncio.BoundedSemaphore(MAX_CONCURRENT_CALLS)
-
-
-def is_retryable(error: Exception) -> bool:
-    """Check if the error is retryable based on defined error types."""
-    return isinstance(error, RETRIABLE_ERROR_TYPES)
-
-
-def log_retry_attempt(error: Exception) -> None:
-    """Log a warning when a retryable error occurs."""
-    logger.warning(f"Retrying due to {error.__class__.__name__} error: {error}")
-
-
-retry_policy = AsyncRetry(
-    predicate=is_retryable,
-    on_error=log_retry_attempt,
-    initial=INITIAL_DELAY_BEFORE_FIRST_RETRY_ATTEMPT,
-    maximum=MAXIMUM_DELAY_BETWEEN_RETRY_ATTEMPTS,
-    multiplier=MULTIPLIER_FOR_EXPONENTIAL_BACKOFF,
-    deadline=TOTAL_TIME_ALLOWED_FOR_RETRIES,
-    timeout=TOTAL_TIME_ALLOWED_FOR_RETRIES,
-)
-
-
-async def execute_with_concurrency_control(
-    callable_func: Callable[..., AsyncGenerator[Any, None]], *args: Any, **kwargs: Any
-) -> ASYNC_GENERATOR_RESYNC_TYPE:
-    """Execute a callable with concurrency control using a bounded semaphore."""
-    async with semaphore:
-        async for item in callable_func(*args, **kwargs):
-            yield item
+from gcp_core.helpers.retry.async_retry import async_generator_retry
+from port_ocean.core.ocean_types import ASYNC_GENERATOR_RESYNC_TYPE
+from aiolimiter import AsyncLimiter
+from typing import Any, Optional
+from loguru import logger
+import asyncio
+
+REQUEST_TIMEOUT = 120.0
+MAXIMUM_CONCURRENT_REQUEST = 100
+semaphore = asyncio.BoundedSemaphore(MAXIMUM_CONCURRENT_REQUEST)
+
+
+@async_generator_retry
+async def paginated_query(
+    client: Any,
+    method: str,
+    request: dict[str, Any],
+    parse_fn: Any,
+    rate_limiter: Optional[AsyncLimiter] = None,
+    *args: Any,
+    **kwargs: Any,
+) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    """
+    General function to handle paginated requests with rate limiting.
+    """
+    page = 0
+    page_token = None
+
+    while True:
+        if page_token:
+            request["page_token"] = page_token
+
+        if rate_limiter:
+            logger.debug(f"Rate limiting enabled for `{method}`")
```

Comment:
add what is the rate limit

Reply:
its logged during the creation of the rate limiter object, but at this stage it will be difficult to get that

Reply:
I agree that this shouldn't be logged here (Since it will mean that for each page this will be printed) but we definitely should log what is the max_rate that we are running with per resource per project.

Add this log to be above the while: true + the AsyncLimiter has a max_rate + time_period properies you can use to print